### PR TITLE
Radiation Solver

### PR DESCRIPTION
**Pull request by owenmylotte**
Moving the radiation solver into the main branch.

### CODE REVIEW

File: src/boundarySolver/boundarySolver.cpp [hunk 1]
```diff
@@ -85,7 +85,7 @@ void ablate::boundarySolver::BoundarySolver::Setup() {
     // Get the geometry for the mesh
     DM faceDM, cellDM;
     VecGetDM(faceGeomVec, &faceDM) >> checkError;
-    VecGetDM(cellGeomVec, &cellDM) >> checkError;
+    VecGetDM(cellGeomVec, &cellDM) >> checkError; //
```
Comment (mmcgurn) [inline src/boundarySolver/boundarySolver.cpp:88]: Do you mind removing this?

File: src/monitors/extractLineMonitor.cpp [hunk 2]
```diff
@@ -33,64 +33,64 @@ void ablate::monitors::ExtractLineMonitor::Register(std::shared_ptr<solver::Solv
...
+    std::vector<double> lineVec; //Vector in the direction of the line
+    for (std::size_t d = 0; d < start.size(); d++) { //
+        L += PetscSqr(end[d] - start[d]); //PetscSqr: squares a number?
+        lineVec.push_back(end[d] - start[d]); //What is push_back? Adds a new element to the vector
     }
-    L = PetscSqrtReal(L);
+    L = PetscSqrtReal(L); //Takes the square root of the sum of differences?
     for (auto& c : lineVec) {
-        c /= L;
+        c /= L; //TODO: What does c represent?
```
Comment (mmcgurn) [inline src/monitors/extractLineMonitor.cpp:56]: TODO: has a meaning in most ides.  Would you mind updating the comments so that we can remove the todo's?

File: src/particles/tracer.cpp [hunk 3]
```diff
@@ -55,7 +55,7 @@ PetscErrorCode ablate::particles::Tracer::freeStreaming(TS ts, PetscReal t, Vec
 
     /* Get local flow velocity */
     const auto &flowVelocityField = particles->subDomain->GetField("velocity");
-    ierr = particles->subDomain->GetFieldLocalVector(flowVelocityField, particles->timeInitial, &vis, &locvel, &vdm);
+    ierr = particles->subDomain->GetFieldLocalVector(flowVelocityField, particles->timeInitial, &vis, &locvel, &vdm); //TODO: Get temperature based on this function
```
Comment (mmcgurn) [inline src/particles/tracer.cpp:58]: I think that we can remove this comment.

File: src/radiation/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+target_sources(ablateLibrary
+        PUBLIC
+        radiation.hpp
+        radiation.cpp
+        )
+
+#add_subdirectory(lodi)
```
Comment (mmcgurn) [inline src/radiation/CMakeLists.txt:7]: Do you mind removing this cmake comments please?

File: src/radiation/radiation.cpp [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+
+void ablate::radiation::RadiationSolver::Setup() {  // allows initialization after the subdomain and dm is established
+    ablate::solver::CellSolver::Setup();
+    dim = subDomain->GetDimensions();  // Number of dimensions already defined in the setup
+
+}
+void ablate::radiation::RadiationSolver::Initialize() {
+    // this->reallySolveParallelPlates();
+    this->RayInit();  // Runs the ray initialization finding cell indices
+    /*this->RayTrace(0);
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:32]: Do you mind if we remove dead code like this?

File: src/radiation/radiation.cpp [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+    for (int i = 1; i < 20; i++) {
+        this->RayProduct(0, i);
+    }*/
+
+}
+
+/// Declaring function for the initialization to call, draws each ray vector and gets all of the cells associated with it (sorted by distance and starting at the boundary working in)
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:39]: These comments are really great and something I should do better with.  Would you mind formatting them as [doxygen](https://www.doxygen.nl/manual/docblocks.html) so that they show up on ablate.dev?

File: src/radiation/radiation.cpp [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+    // March over each cell in this region to create the stencil
+    // IS cellIS;
+    // PetscInt cStart, cEnd;  // Keep these as local variables, the cell indices in the radiation initializer should be updated with every radiation initialization. The variables here are owned by the
+                            // DM
+    stencilSet.clear(); //Clear the existing set because a new set of cells will be created below
+
+    solver::Range cellRange;
+    GetCellRange(cellRange);
+    // std::set<PetscInt> stencilSet; //Moved to a class variable
+    for (PetscInt c = cellRange.start; c < cellRange.end; ++c) {
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:56]: Do we need the stencilSet or could we march directly over the cellRange instead?

File: src/radiation/radiation.cpp [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+    DM cellDM;
+    VecGetDM(cellGeomVec, &cellDM);
+    DMPlexGetGeometryFVM(cellDM, nullptr, &cellGeomVec, &minCellRadius);  // Obtain the geometric information about the cells in the DM?
+    VecGetArrayRead(cellGeomVec, &cellGeomArray);
+    PetscFVCellGeom* cellGeom;
+
+    PetscInt ncells = 0;
+    for (auto iCell : stencilSet) {
+        // double percentComplete = 100.0 * double(ncells) / double(stencilSet.size());
+        // PetscPrintf(PETSC_COMM_WORLD, "Radiation Initializer Percent Complete: %f\n", percentComplete);
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:83]: I think that we should use a [ablate::monitors::logs::Log](https://ablate.dev/content/simulations/components/ablate::monitors::logs::Log.html) for progress variable.  That way by default there is not log so nothing reports, but we could provide a log if we want to report the status.

File: src/radiation/radiation.cpp [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+                    // PetscPrintf(PETSC_COMM_WORLD, "Intensity: %f\n", rayIntensity);
+                }
+                /// The rays end here, their intensity is added to the total intensity of the cell
+                intensity += rayIntensity * sin(theta) * dTheta * dPhi;  // Gives the partial impact of the ray on the total sphere. The sin(theta) is a result of the polar coordinate discretization
+            }
+            // PetscPrintf(PETSC_COMM_WORLD, "Radiative Gain: %g\n", intensity);
+        }
+
+        //Total energy gain of the current cell depends on absorptivity at the current cell
+        PetscPrintf(PETSC_COMM_WORLD, "Radiative Gain: %g\n", intensity);
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:253]: I think we may want this to output to the same log earlier.

File: src/radiation/radiation.cpp [hunk 10]
```diff
@@ -0,0 +1,621 @@
...
+        stream << "\n";
+
+        ncells++;
+        // Cleanup
+        VecRestoreArray(loctemp, &temperatureArray);
+    }
+    stream.close();
+}
+
+PetscReal ablate::radiation::RadiationSolver::ReallySolveParallelPlates(PetscReal z) {  ///Analytical solution of a special verification case.
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:265]: Where is this used?  If it is only in the test can we move it to the test file?

File: src/radiation/radiation.cpp [hunk 11]
```diff
@@ -0,0 +1,621 @@
...
+}
+
+/// End of the added radiation stuff
+PetscErrorCode ablate::radiation::RadiationSolver::ComputeRHSFunction(PetscReal time, Vec locXVec,
+                                                                      Vec rhs) {  // main interface for integrating in time Inputs: local vector, x vector (current solution), local f vector
+    PetscFunctionBeginUser;                                                           // gets fields out of the main vector
+
+    /// Get setup things for the position vector of the current cell index
+    /* const auto& errorField = subDomain->GetField("error");
+    PetscScalar* errorArray = nullptr;
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:513]: Can we kill this unused code?

File: src/solver/solver.hpp [hunk 12]
```diff
@@ -43,6 +43,7 @@ class Solver {
     static PetscErrorCode DMPlexInsertTimeDerivativeBoundaryValues_Plex(DM dm, PetscDS ds, PetscBool insertEssential, Vec locX, PetscReal time, Vec faceGeomFVM, Vec cellGeomFVM, Vec gradFVM);
 
    public:
+    Solver();
```
Comment (mmcgurn) [inline src/solver/solver.hpp:46]: Why was this needed?

File: src/radiation/radiation.hpp [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+namespace ablate::radiation {
+
+class RadiationSolver : public solver::CellSolver, public solver::RHSFunction {  //Cell solver provides cell based functionality, right hand side function compatibility with finite element/ volume
+   public:
+    PetscInt raynumber;
+
+    /**
+     * Boundary information.
+     */
+    typedef struct {
```
Comment (mmcgurn) [inline src/radiation/radiation.hpp:21]: Are you using this struct?

File: src/radiation/radiation.hpp [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    static PetscReal ReallySolveParallelPlates(PetscReal z); //std::vector<PetscReal>
+    static PetscReal EInteg(int order, double x);
+
+    ///Class Constants
+    const PetscReal sbc = 5.6696e-8;  // Stefan-Boltzman Constant (J/K)
+    const PetscReal refTemp = 298.15;
+    const PetscReal pi = 3.1415926535897932384626433832795028841971693993;
+
+    ///Class inputs and Variables
+    DM vdm; //Abstract PETSc object that manages an abstract grid object and its interactions with the algebraic solvers
```
Comment (mmcgurn) [inline src/radiation/radiation.hpp:83]: Do these need to live in the class definition or can be local variables in each function?

File: src/radiation/radiation.hpp [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * @param ctx
+     * @return
+     */
+    PetscErrorCode ComputeRHSFunction(PetscReal time, Vec locXVec, Vec locFVec) override;
+
+    /**
+     * Helper function to project values to a cell boundary instead of the cell centroid
+     */
+    void InsertFieldFunctions(const std::vector<std::shared_ptr<mathFunctions::FieldFunction>>& initialization, PetscReal time = 0.0);
```
Comment (mmcgurn) [inline src/radiation/radiation.hpp:57]: Are these functions still used for the erudition solver?

File: src/radiation/radiation.hpp [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * Return a reference to the boundary geometry.  This is a slow call and should only be done for init/debugging/testing
+     */
+    const BoundaryFVFaceGeom& GetBoundaryGeometry(PetscInt cell) const;
+
+    ///Starting added radiation stuff here
+
+    ///Class Methods
+    void RayTrace(PetscReal time);
```
Comment (mmcgurn) [inline src/radiation/radiation.hpp:67]: If these methods are only used by the class can we put them in a private section?

File: tests/integrationTests/inputs/shocktube/shockTube2Gas2D.yaml [hunk 17]
```diff
@@ -48,32 +48,39 @@ timestepper:
...
-      fluxCalculatorGasLiquid: !ablate::finiteVolume::fluxCalculator::Riemann2Gas
-        eosL: *4
-        eosR: *1
-      fluxCalculatorLiquidGas: !ablate::finiteVolume::fluxCalculator::Riemann2Gas
-        eosL: *1
-        eosR: *4
-      fluxCalculatorLiquidLiquid: !ablate::finiteVolume::fluxCalculator::Riemann2Gas
-        eosL: *1
-        eosR: *1
+solvers:
```
Comment (mmcgurn) [inline tests/integrationTests/inputs/shocktube/shockTube2Gas2D.yaml:51]: Can be move the radiation solver to a new integration test?

File: tests/integrationTests/inputs/geom/slabBurner.v1.stp [hunk 18]
```diff
@@ -0,0 +1,289 @@
+ISO-10303-21;
```
Comment (mmcgurn) [inline tests/integrationTests/inputs/geom/slabBurner.v1.stp:1]: I don't think we are using the slabBurner geom in the repo, can we remove this file?

**Review by mmcgurn — CHANGES_REQUESTED** [review]
@owenmylotte This is some great work!  I have some requests about code cleanup and organization if you don't mind taking a look.  Please feel free to stop by and we can look at the changes together.

File: src/radiation/radiation.hpp (hunk 13 above)
Comment (owenmylotte) [reply src/radiation/radiation.hpp:21]: I thought that this is where the cell center location is accessed from. If that isn't the case then I wouldn't need it.

File: src/radiation/radiation.hpp (hunk 14 above)
Comment (owenmylotte) [reply src/radiation/radiation.hpp]: I've moved the temperature field objects to be local since they only come up in `ComputeRHSFunction`. The other variables are accessed by multiple functions or come in through the class constructor. Should these variables be handled some other way?

File: src/radiation/radiation.cpp [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+     * (sorted by distance and starting at the boundary working in)
+     * */
+
+    double theta;  //!< represents the actual current angle (inclination)
+    double phi;    //!< represents the actual current angle (rotation)
+
+    /**Locally get a range of cells that are included in this subdomain at this time step for the ray initialization
+     * */
+    solver::Range cellRange;
+    GetCellRange(cellRange);
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:46]: You will need to restore the cell range after you use it.

File: src/radiation/radiation.cpp [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+    /** Get setup things for the position vector of the current cell index
+     * Declare the variables that will contain the geometry of the cells
+     * Obtain the geometric information about the cells in the DM
+     * */
+    const PetscScalar* cellGeomArray;
+    PetscReal minCellRadius;
+    DM cellDM;
+    VecGetDM(cellGeomVec, &cellDM);
+    DMPlexGetGeometryFVM(cellDM, nullptr, &cellGeomVec, &minCellRadius);
+    VecGetArrayRead(cellGeomVec, &cellGeomArray);
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:67]: This needs to be restored

**Review by mmcgurn — CHANGES_REQUESTED** [review]
